docs/architecture/017-applicant-ux.md

```diff
@@ -0,0 +1,86 @@
+# 017: Applicant Form UX
+
+Authors: @opqdonut, @Macroz, @marharyta
+
+This ADR aims to collect together some topics like validations in
+frontend, autosaving, etc. that are related to the Applicant's user
+experience. There have been previous discussions about all of these
+topics but no unified presentation.
+
+Based on ticket [#2639](https://github.com/CSCfi/rems/issues/2639)
+
+# Problems with the current applicant UX
+
+## A) Validation errors block saving
+
+You don't know if your inputs (e.g. ip address) are valid before
+pressing Save, and then you need to fix all of these errors before
+REMS saves your work.
+
+This can be irritating if you're working on a long application, and
+just want to save your work and go home. It's less of an issue if
+users save early & often.
+
+**Possible solution**: Show validation errors to the user, but perform
+the save anyway. Show errors in yellow?
+
+Historical note: historically (in the pre-SPA days) REMS used to work
+like this. This topic was more recently touched on in
+[#2117](https://github.com/CSCfi/rems/issues/2117). Interestingly
+enough, the idea idea seems to have been to only perform schema-style
+validations on drafts. However, the email content validation was
+included in the spec, so subsequent field types added their content
+validations to save-draft (see e.g.
+[#2567](https://github.com/CSCfi/rems/pull/2567),
+[#2594](https://github.com/CSCfi/rems/pull/2594))
+
+
+## B) Validation errors are shown late
+
+You need to perform an action (save/submit) to see the validation
+errors. Technically we should be able to run the validation code on
+the frontend at any point. There's no need to talk to the backend.
+
+**Possible solution**: Validate when user stops typing / leaves field.
+
+Questions:
+- What about partially filled in forms? It would be best to not show validation errors for fields that haven't been edited yet.
```

Comment:
If the field is empty, is makes sense to throw isEmpty validation on submit/save, if it is filled with data, it makes sense to validate the data

Reply:
Let's say if you fill two out of three fields and save, I would not like to show an error of not having filled the field three yet (because you did not proceed that far yet). Same but stronger if there exist 30 fields and you just filled the first.